### DISCUSSION TITLE
How to setup with nginx, with frontend included?

### DISCUSSION THREAD

**Discussion by ron0studios** [pocketbase/pocketbase#2266] [Q&A]
Hi,
    For the past day or so, I've been trying to setup pocketbase on another server, through ssh, using nginx. So far the frontend has already been set up, and it runs in port 80.
I looked into the pocketbase documentation on how to do this myself and have run into several issues, and am very confused. 
- Firstly, I tried simply running the command `./pocketbase serve --http="<IP HERE>:80"` and got the error `listen tcp <IP HERE>: bind: cannot assign requested address`. I tried, `sudo` too, which also did not work.
- Then, I tried assigning it to the same port and domain from which my frontend is listening to, which was `0.0.0.0:80`, and, well, I got the error `bind: address already in use`.
- I then tried a different port `81`, which worked when running the executable. I then change from `80` to `81` after copy and pasting the nginx config from the pocketbase docs. But now I'm getting a CORS error.

This is definitely not a fault of pocketbase, but rather my confusion, so I decided to post it here instead of as an issue. 

Could anyone please kindly help to guide me through a more step by step installation of pocketbase with a SPA (e.g Vue) on nginx? 

I have not set up https, but I have checked the CORS requests to see that both source and destination are http. 
I am running an oracle cloud server.

Thank you!

**Comment by ganigeorgiev** [comment]
If you already have nginx listening on 80/443 ports then you can't bind another process to those ports (_unless `SO_REUSEPORT` or similar is used_).

So in this case you have to run PocketBase on some other local port eg. `8090` and use the `proxy_pass` nginx directive as shown in https://pocketbase.io/docs/going-to-production/#using-reverse-proxy.

PocketBase by default allows all origins (_unless you've started the executable with explicit `--origins` flag_) so the CORS error is most likely caused by some misconfiguration in the frontend app but there aren't enough details how the frontend app is deployed and configured to help you debug it.

Note that if PocketBase is the only application on the server and your frontend app is entirely client-side (aka. you are not using node server-side rendering), then you don't need nginx or anything else. You can put your static files inside a `pb_public` directory and PocketBase will take care to serve them. In this case you can follow https://pocketbase.io/docs/going-to-production/#minimal-setup.

**Reply by ron0studios** [reply]
thanks for the help! So if i run `sudo ./pocketbase serve` without the `--http` would it work? For the config, should i be listening to the new port i.e 8090? When i do that, the issue is that the address already becomes in use, so I can't use both nginx and pocketbase on the same port! This is all very confusing...

**Reply by ron0studios** [reply]
i also don't have a domain name and am working on ip addresses

**Reply by ron0studios** [reply]
UPDATE:
so ive done some thinking and concluded that it must be an issue on the client side (since it cant be an issue with the pocketbase executable). 
How can you enable cors in the pocketbase javascript sdk?

**Reply by ron0studios** [reply]
UPDATE 2: 
i tried running this on every instance in the javascript sdk, to no avail.
```

pb.beforeSend = function (url, options) {
  // For list of the possible request options properties check
  // https://developer.mozilla.org/en-US/docs/Web/API/fetch#options
  options.headers = Object.assign({}, options.headers, {
      'Access-Control-Allow-Origin': '*',
  });

  return { url, options };
};
```
I am so stuck and have no idea what to do.

**Reply by ron0studios** [reply]
![image](https://user-images.githubusercontent.com/47331292/231272514-2366d6db-df7e-40bb-9403-7dd7c8e58d3c.png)
only the OPTIONS request is getting the cors error and not the GET request. I don't remember calling an OPTIONS request ever?

@ganigeorgiev if you're available, your help would be much appreciated, thank you!!!

**Reply by ganigeorgiev** [reply]
@ron0studios I don't know how to help you. You are spamming the discussion with random replies and changes in your code but without providing any context or general information about your setup so that I can try to reproduce it.

I don't understand why you are sending `Access-Control-Allow-Origin` header in the `beforeSend` hook and for me it doesn't make sense.

If you are still not able to resolve the issue, please try to provide:

- general information about your environment
- share your nginx config
- how do you deploy PocketBase?
- how do you deploy your frontend app?
   - do you have a node server running? And if Yes, why if your frontend app is a SPA?
- (_ideally_) **minimal reproducible** repo

**Comment by TMBL-DEV** [comment]
This is just a random guess cuz there is no code accessible. But make sure you allow nginx cors requests. By putting this in your nginx config


server{
   ...
   add_header Access-Control-Allow-Origin *;
   ...
}

And if that doesn't work look at this. It might help you understand a bit more.

https://youtu.be/4KHiSt0oLJ0
